packages/vite/src/node/server/index.ts

```diff
@@ -469,22 +465,26 @@ export async function _createServer(
     setClientErrorHandler(httpServer, config.logger)
   }
 
-  // eslint-disable-next-line eqeqeq
-  const watchEnabled = serverConfig.watch !== null
-  const watcher = watchEnabled
-    ? (chokidar.watch(
-        // config file dependencies and env file might be outside of root
-        [
-          root,
-          ...config.configFileDependencies,
-          ...getEnvFilesForMode(config.mode, config.envDir),
-          // Watch the public directory explicitly because it might be outside
-          // of the root directory.
-          ...(publicDir && publicFiles ? [publicDir] : []),
-        ],
-        resolvedWatchOptions,
-      ) as FSWatcher)
-    : createNoopWatcher(resolvedWatchOptions)
```

Comment:
I removed `NoopWatcher` in favor of patching the chokidar watcher with `watcher.add = function () {  return this }` as `FSWatcher.options` now return a resolved option (non-nullable) which we'd have to copy over from chokidar if we retain `NoopWatcher`.

It was also initially hard to make types happy because `chokidar` export types for `FSWatcher` contained a bunch of internal properties making `implements` unhappy. 

This trick could be easier to maintain in the long run instead.

Reply:
but doesn't it start watching when you call `watch`? 🤔

Reply:
In resolveWatchOptions I added a config to ignore every path if this is null, so no fs.watch will happen but there will be a chokidar fs watcher instance that does nothing.

Reply:
It was also possible to do with the previous watcher, but I went with a noop watcher specifically to avoid having a running process. Even if every path is ignored, it will still try to check them and have an expensive watcher do nothing.

Reply:
I don't know if it's exactly expensive. The alternative like before is harder to implement if we want to stick with it.

Reply:
> The alternative like before is harder to implement if we want to stick with it.

How hard can it be to implement a noop? 

> I don't know if it's exactly expensive

Would you call `fs.watch` in your every project just for lols? Why start a watcher to ignore it?

Reply:
As I mentioned before, the ignore will already block any fs.watch from happening in the first place. Happy to be proven wrong here.

Reply:
> How hard can it be to implement a noop?

Ok, I checked - and the only thing that changed is the options - Is it really worth it to start a new watcher every time `createServer` is called just to avoid adding mocked options?

Reply:
> As I mentioned before, the ignore will already block any fs.watch from happening in the first place. Happy to be proven wrong here.

I don't see any filtering happening to files passed down to `add` which is called when `chockidr.watch(paths)` is called: 

https://github.com/paulmillr/chokidar/blob/97894d36657c8e4571bd1a07fb6e2a3f50f7019d/src/index.ts#L442

Seems like ignoring happens when it's already watching. But maybe using `.watch([])` will work better.

Reply:
We can also close it immediately maybe

Reply:
I've added more guards here: https://github.com/vitejs/vite/pull/18521